internal/cloudsql/refresh_test.go

```diff
@@ -15,77 +15,334 @@
+	}
+	if wantPrivateIP != gotIP {
+		t.Fatalf("metadata IP mismatch, want = %v, got = %v", wantPrivateIP, gotIP)
+	}
+	if wantExpiry != gotExpiry {
+		t.Fatalf("expiry mismatch, want = %v, got = %v", wantExpiry, gotExpiry)
+	}
+	if wantConnName != tlsCfg.ServerName {
+		t.Fatalf("server name mismatch, want = %v, got = %v", wantConnName, tlsCfg.ServerName)
 	}
 }
-func TestFetchEphemeralCert(t *testing.T) {
-	ctx := context.Background()
 
-	// define some test instance settings
-	cn, err := parseConnName("my-proj:my-region:my-inst")
+func TestRefreshFailsFast(t *testing.T) {
+	cn, _ := parseConnName("my-project:my-region:my-instance")
+	inst := mock.NewFakeCSQLInstance("my-project", "my-region", "my-instance")
+	client, cleanup, err := mock.NewSQLAdminService(
+		context.Background(),
+		mock.InstanceGetSuccess(inst, 1),
+		mock.CreateEphemeralSuccess(inst, 1),
+	)
 	if err != nil {
-		t.Fatalf("%s", err)
+		t.Fatalf("failed to create test SQL admin service: %s", err)
 	}
+	defer cleanup()
+
+	r := newRefresher(time.Hour, 30*time.Second, 1, client)
+	_, _, _, err = r.performRefresh(context.Background(), cn, RSAKey)
+	if err != nil {
+		t.Fatalf("expected no error, got = %v", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+	// context is canceled
+	_, _, _, err = r.performRefresh(ctx, cn, RSAKey)
+	if !errors.Is(err, context.Canceled) {
+		t.Fatalf("expected context.Canceled error, got = %v", err)
+	}
+
+	// force the rate limiter to throttle with a timed out context
+	ctx, _ = context.WithTimeout(context.Background(), time.Millisecond)
+	_, _, _, err = r.performRefresh(ctx, cn, RSAKey)
+
+	if !errorContains(err, "throttled") {
+		t.Fatalf("expected throttled error, got = %v", err)
+	}
+}
+
+func TestRefreshWithFailedMetadataCall(t *testing.T) {
+	cn, _ := parseConnName("my-project:my-region:my-instance")
+
+	testCases := []struct {
+		req     *mock.Request
+		wantErr string
+		desc    string
+	}{
+		{
+			req: mock.CreateEphemeralSuccess(
+				mock.NewFakeCSQLInstance(cn.project, cn.region, cn.name), 1),
+			wantErr: "failed to get instance",
+			desc:    "When the Metadata call fails",
+		},
+		{
+			req: mock.InstanceGetSuccess(
+				mock.NewFakeCSQLInstance(cn.project, cn.region, cn.name,
+					mock.WithRegion("some-other-region")), 1),
+			wantErr: "region was mismatched",
+			desc:    "When the region does not match",
+		},
+		{
+			req: mock.InstanceGetSuccess(
+				mock.NewFakeCSQLInstance(
+					cn.project, cn.region, cn.name,
+					mock.WithRegion("my-region"),
+					mock.WithFirstGenBackend(),
+				), 1),
+			wantErr: "only Second Generation",
+			desc:    "When the instance isn't Second generation",
+		},
+		{
+			req: mock.InstanceGetSuccess(
+				mock.NewFakeCSQLInstance(
+					cn.project, cn.region, cn.name,
+					mock.WithRegion("my-region"),
+					mock.WithMissingIPAddrs(),
+				), 1),
+			wantErr: "no supported IP addresses",
+			desc:    "When the instance has no supported IP addresses",
```

Comment:
Should this be no supported? Or should it be none of the type specified?

Reply:
I have this in here to cover these lines:

https://github.com/kurtisvg/cloud-sql-go-connector/blob/cab162906d2bec0862d86a76af1bff2d42d3b2f3/internal/cloudsql/refresh.go#L73-L75

We could add a function option to configure a custom IP type, e.g., `mock.WithIPType("foo")`. Perhaps this would be useful for future work? WDYT?

Reply:
Ah OK - this seems fine.